File: app/src/main/kotlin/com/example/unittesting/login/presenter/LoginPresenter.java
package com.example.unittesting.login.presenter;

import com.example.unittesting.R;
import com.example.unittesting.ResourceProvider;
import com.example.unittesting.SchedulersFactory;
import com.example.unittesting.BasePresenter;
import com.example.unittesting.login.model.LoginCredentials;
import com.example.unittesting.login.model.LoginUseCase;
import com.example.unittesting.login.model.LoginValidator;

import io.reactivex.functions.Consumer;

public class LoginPresenter extends BasePresenter<LoginView> {

    ResourceProvider resourceProvider;
    LoginValidator loginValidator;
    LoginUseCase loginUseCase;
    SchedulersFactory schedulersFactory;

    public LoginPresenter(ResourceProvider resourceProvider, LoginValidator loginValidator, LoginUseCase loginUseCase, SchedulersFactory schedulersFactory) {
        this.resourceProvider = resourceProvider;
        this.loginValidator = loginValidator;
        this.loginUseCase = loginUseCase;
        this.schedulersFactory = schedulersFactory;
    }

    public void attemptLogin(LoginCredentials loginCredentials) {

        boolean validationError = validatePassword(loginCredentials, false);

        validationError = validateLogin(loginCredentials, validationError);

        if (validationError) {
            return;
        }

        getView().showProgress();

        loginUseCase.loginWithCredentialsWithStatus(loginCredentials)
                .compose(schedulersFactory.<Boolean>createMainThreadSchedulerTransformer())
                .subscribe(new Consumer<Boolean>() {

                    @Override
                    public void accept(Boolean success) throws Exception {
                        getView().hideProgress();

                        if (success) {
                            getView().onLoginSuccessful();
                        } else {
                            getView().showPasswordError(resourceProvider.getString(R.string.error_incorrect_password));
                            getView().requestPasswordFocus();
                        }
                    }
                });
    }

    private boolean validatePassword(LoginCredentials loginCredentials, boolean validationError) {
        if (!loginValidator.validatePassword(loginCredentials.password)) {
            getView().showPasswordError(resourceProvider.getString(R.string.error_invalid_password));
            getView().requestPasswordFocus();
            validationError = true;
        } else {
            getView().showPasswordError(null);
        }
        return validationError;
    }

    private boolean validateLogin(LoginCredentials loginCredentials, boolean validationError) {
        if (!loginValidator.validateLogin(loginCredentials.login)) {
            getView().showLoginError(resourceProvider.getString(R.string.error_field_required));
            getView().requestLoginFocus();
            validationError = true;
        } else {
            getView().showLoginError(null);
        }
        return validationError;
    }
}

File: app/src/main/kotlin/com/example/unittesting/login/presenter/LoginPresenter.kt
package com.example.unittesting.login.presenter

import com.example.unittesting.BasePresenter
import com.example.unittesting.R
import com.example.unittesting.ResourceProvider
import com.example.unittesting.SchedulersFactory
import com.example.unittesting.login.model.LoginCredentials
import com.example.unittesting.login.model.LoginUseCase
import com.example.unittesting.login.model.LoginValidator

class LoginPresenter(private val resourceProvider: ResourceProvider, private val loginValidator: LoginValidator, private val loginUseCase: LoginUseCase, private val schedulersFactory: SchedulersFactory) : BasePresenter<LoginView>() {

    fun attemptLogin(loginCredentials: LoginCredentials) {
        if (!validateInputs(loginCredentials)) {
            return
        }

        view?.showProgress()
        loginUseCase.loginWithCredentialsWithStatus(loginCredentials)
                .compose(schedulersFactory.createMainThreadSchedulerTransformer())
                .subscribe { success ->
                    view?.hideProgress()

                    if (success) {
                        view?.onLoginSuccessful()
                    } else {
                        view?.showPasswordError(resourceProvider.getString(R.string.error_incorrect_password))
                        view?.requestPasswordFocus()
                    }
                }
    }

    private fun validateInputs(loginCredentials: LoginCredentials): Boolean {
        val validateLogin = validateLogin(loginCredentials)
        return validatePassword(loginCredentials) && validateLogin //XXX validateLogin is not inlined to avoid short circuit check
    }

    private fun validateLogin(loginCredentials: LoginCredentials): Boolean =
            if (loginValidator.validateLogin(loginCredentials.login)) {
                view?.showLoginError(null)
                true
            } else {
                view?.showLoginError(resourceProvider.getString(R.string.error_field_required))
                view?.requestLoginFocus()
                false
            }

    private fun validatePassword(loginCredentials: LoginCredentials): Boolean =
            if (loginValidator.validatePassword(loginCredentials.password)) {
                view?.showPasswordError(null)
                true
            } else {
                view?.showPasswordError(resourceProvider.getString(R.string.error_invalid_password))
                view?.requestPasswordFocus()
                false
            }
}
